### DISCUSSION TITLE
TS2707 [ERROR]: Generic type 'RouterContext<R, P, S>' requires between 1 and 3 type arguments

### DISCUSSION THREAD

**Discussion by finlaydotb** [oakserver/oak#468] [Q&A] [answered]
Upgrading from version `6.5.0` to `10.2.0` now give the following compile error:

```
TS2707 [ERROR]: Generic type 'RouterContext<R, P, S>' requires between 1 and 3 type arguments
```

so for example the following code does not compile anymore:

```
router.get("/", isAuthed, indexHandler)
router.get("/index", isAuthed, indexHandler)
```

```
export const indexHandler = async (ctx: RouterContext) => {
...
}
```

I could not find any thing in the documentation about what changes with `RouterContext` and how to use it now.

**Comment by kitsonk** [comment, marked as answer]
The change was mentioned as a breaking change in [v10.0.0](https://github.com/oakserver/oak/blob/main/CHANGELOG.md#version-1000).

The error is accurate, in that 1 to 3 generic parameters are required.  The first generic parameter is the route, which is then used to infer the route parameters.  When the middleware is expressed in line, TypeScript can infer the route from the `.get()`, `.all()` etc. methods.  In your case though, you need to supply it, because it cannot be inferred:

```ts
export const indexHandler = async (ctx: RouterContext<"/" | "/index">) => {
  //
}
```

That should resolve your issue.

**Reply by youngyou** [reply]
What about this code? 
```
router.all(
  "/foo/:path*",
  async (ctx: RouterContext<"/foo/:path*", { path: string }, AuthState>) => {
    // more code
  }
);
```
This will throw 
```
Type '{ path: string; }' is not assignable to type '{ "path*": string; } & Record<string | number, string | undefined>'
```
